File: src/class037/C7_HouseRobberIII.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Code07_HouseRobberIII:
    def rob(self, root):
        res = self.f(root)
        return max(res[0], res[1])

    def f(self, root):
        if root is None:
            return (0, 0)
        left = self.f(root.left)
        right = self.f(root.right)
        # 当前节点选择偷，下一个就不能偷
        yes = root.val + left[1] + right[1]
        # 当前节点选择不偷，下一个可以偷或不偷，选择一个较大的
        no = max(left[0], left[1]) + max(right[0], right[1])
        return (yes, no)

# 使用中文注释来描述代码的功能和逻辑

File: src/class037/C7_HouseRobberIII.cpp
#include <iostream>
#include <algorithm>
using namespace std;

// 包裹类 class037
namespace class037 {

// 二叉树打家劫舍问题
// 测试链接 : https://leetcode.cn/problems/house-robber-iii/
class Code07_HouseRobberIII {
public:
    struct TreeNode {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    };

    // 提交以下方法
    static int rob(TreeNode* root) {
        f(root);
        return max(yes, no);
    }

private:
    // 全局变量，完成了X子树的遍历，返回之后
    // yes变成，X子树在偷头节点的情况下，最大的收益
    static int yes;

    // 全局变量，完成了X子树的遍历，返回之后
    // no变成，X子树在不偷头节点的情况下，最大的收益
    static int no;

    static void f(TreeNode* root) {
        if (root == nullptr) {
            yes = 0;
            no = 0;
        } else {
            int y = root->val;
            int n = 0;
            f(root->left);
            y += no;
            n += max(yes, no);
            f(root->right);
            y += no;
            n += max(yes, no);
            yes = y;
            no = n;
        }
    }
};

// Initialize static members
int Code07_HouseRobberIII::yes = 0;
int Code07_HouseRobberIII::no = 0;

} // namespace class037
